### ISSUE TITLE
Python2 vs. Python3 for running generator

### ISSUE THREAD

**Issue by mlo-poc** [galou/inkscape_generator#3]
Hi,

first it seemd, there **was** a dependency besides the ones, that inkscape itself has:

```
File "...extensions/generator.py", line 51, in <module>
    import StringIO
ModuleNotFoundError: No module named 'StringIO'
```

From https://stackoverflow.com/questions/11914472/stringio-in-python3 i learned:
> From What’s New In Python 3.0:
>
>    The StringIO and cStringIO modules are gone. Instead, import the io module and use io.StringIO or io.BytesIO for text and data respectively.

So the reason for this is *me using a python3-only setup*.

---

I tried fixing the first one and immediately hit the next error:

```
Traceback (most recent call last):
  File "/home/mlo/.config/inkscape/extensions/generator.py", line 336, in <module>
    e.run()
  File "/usr/share/inkscape/extensions/inkex/base.py", line 140, in run
    self.save_raw(self.effect())
  File "/home/mlo/.config/inkscape/extensions/generator.py", line 109, in effect
    self.handle_csv()
  File "/home/mlo/.config/inkscape/extensions/generator.py", line 128, in handle_csv
    self.header = reader.next()
AttributeError: '_csv.reader' object has no attribute 'next'
sys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name='/home/mlo/Dokumente/20_layout/10_ostercappeln/32_urkunden/2021/konfirmanden_daten_oelbaeume.csv' mode='r' encoding='UTF-8'>

```
Again stackexchange was helpful: https://stackoverflow.com/questions/42767250/python-csv2libsvm-py-attributeerror-csv-reader-object-has-no-attribute-nex
> This is because of the differences between python 2 and python 3. Use the built-in function next in python 3. That is, write next(reader) instead of reader.next() in line 47. In addition, you should open the file in the text mode. So, change line 47 as i = open( input_file, 'r' ).

ok...
I went on and soon found myself porting your script to python3.

**Comment by mlo-poc** [comment]
Here's my feable attempt to port the generator to python3
I introduced a lot of redundant code, so some cleaning up might be needed.

```
diff --git a/generator.py b/generator.py
index 342ef9e..619581d 100755
--- a/generator.py
+++ b/generator.py
@@ -43,12 +43,16 @@ Release notes:
 
 from __future__ import division
 
+import sys
 import os
 import tempfile
 import csv
 import re
 import shutil
-import StringIO
+try:
+    from StringIO import StringIO ## for Python 2
+except ImportError:
+    from io import StringIO ## for Python 3
 from lxml import etree
 from xml.sax.saxutils import escape
 import inkex
@@ -122,7 +126,10 @@ class Generator(inkex.Effect):
             raise Exception(_('Cannot read "{}"'.format(self.options.datafile)))
         if self.options.var_type == 'name':
             try:
-                self.header = reader.next()
+                if sys.version_info[0] == 3:
+                    self.header = next(reader)
+                else:
+                    self.header = reader.next()
             except StopIteration:
                 errormsg(_('Data file "{}" contains no data'.format(
                     self.options.datafile)))
@@ -145,7 +152,10 @@ class Generator(inkex.Effect):
 
     def create_svg(self, name_dict):
         """Writes out a modified svg"""
-        s = StringIO.StringIO()
+        if sys.version_info[0] == 3:
+            s = StringIO()
+        else:
+            s = StringIO.StringIO()
         for svg_line in open(self.options.input_file, 'r').readlines():
             # Modify the line to handle replacements from extension GUI
             svg_line = self.expand_extra_vars(svg_line, name_dict)
@@ -153,14 +163,14 @@ class Generator(inkex.Effect):
             svg_line = self.expand_vars(svg_line, name_dict)
             s.write(svg_line)
         # Modify the svg to include or exclude groups
-        root = etree.fromstring(s.getvalue())
+        root = etree.fromstring(bytes(s.getvalue(), encoding='utf-8'))
         self.filter_layers(root, name_dict)
         svgout = self.get_svgout()
         try:
             f = open(svgout, 'w')
             f.write(etree.tostring(root,
                                    encoding='utf-8',
-                                   xml_declaration=True))
+                                   xml_declaration=True).decode('utf-8'))
         except IOError:
             errormsg(_('Cannot open "' + svgout + '" for writing'))
         finally:
@@ -212,8 +222,13 @@ class Generator(inkex.Effect):
         """Replace %VAR_???% with the content from a csv entry"""
         if '%' not in line:
             return line
-        for k, v in name_dict.iteritems():
-            line = line.replace('%VAR_' + k + '%', escape(v))
+            # Python 3 renamed dict.iteritems -> dict.items. – Blender May 23 '15 at 23:16
+        if sys.version_info[0] == 3:
+            for k, v in name_dict.items():
+                line = line.replace('%VAR_' + k + '%', escape(v))
+        else:
+            for k, v in name_dict.iteritems():
+                line = line.replace('%VAR_' + k + '%', escape(v))
         return line
 
     def filter_layers(self, root, name_dict):
@@ -286,24 +301,43 @@ class Generator(inkex.Effect):
                         + '--export-type=' + fmt + ' '
                         + '--export-filename="' + outfile + '" '
                         '"' + svgfile + '"')
-
-        for line, svgfile in self.svgouts.iteritems():
-            d = self.get_line_desc(line)
-            outfile = self.get_output(d)
-            if self.options.format == 'jpg':
-                # TODO: output a jpg file
-                self.options.format = 'png'
-                outfile = outfile.replace('jpg', 'png')
-            if self.options.format == 'svg':
-                try:
-                    shutil.move(svgfile, outfile)
-                except OSError:
-                    errormsg(_('Cannot create "' + outfile + '"'))
-            else:
-                cmd = get_export_cmd(svgfile,
-                                     self.options.format,
-                                     self.options.dpi, outfile)
-                os.system(cmd)
+            # Python 3 renamed dict.iteritems -> dict.items. – Blender May 23 '15 at 23:16
+        if sys.version_info[0] == 3:
+            for line, svgfile in self.svgouts.items():
+                d = self.get_line_desc(line)
+                outfile = self.get_output(d)
+                if self.options.format == 'jpg':
+                    # TODO: output a jpg file
+                    self.options.format = 'png'
+                    outfile = outfile.replace('jpg', 'png')
+                if self.options.format == 'svg':
+                    try:
+                        shutil.move(svgfile, outfile)
+                    except OSError:
+                        errormsg(_('Cannot create "' + outfile + '"'))
+                else:
+                    cmd = get_export_cmd(svgfile,
+                                         self.options.format,
+                                         self.options.dpi, outfile)
+                    os.system(cmd)
+        else:
+            for line, svgfile in self.svgouts.iteritems():
+                d = self.get_line_desc(line)
+                outfile = self.get_output(d)
+                if self.options.format == 'jpg':
+                    # TODO: output a jpg file
+                    self.options.format = 'png'
+                    outfile = outfile.replace('jpg', 'png')
+                if self.options.format == 'svg':
+                    try:
+                        shutil.move(svgfile, outfile)
+                    except OSError:
+                        errormsg(_('Cannot create "' + outfile + '"'))
+                else:
+                    cmd = get_export_cmd(svgfile,
+                                         self.options.format,
+                                         self.options.dpi, outfile)
+                    os.system(cmd)
 
     def show_preview(self):
         systems = {
@@ -324,8 +358,12 @@ class Generator(inkex.Effect):
     def clean(self):
         """Delete temporary svg files and directory"""
         if self.options.format != 'svg':
-            for svgfile in self.svgouts.itervalues():
-                os.remove(svgfile)
+            if sys.version_info[0] == 3:
+                for svgfile in self.svgouts.values():
+                    os.remove(svgfile)
+            else:
+                for svgfile in self.svgouts.itervalues():
+                    os.remove(svgfile)
         os.rmdir(self.tmpdir)
 
 if __name__ == '__main__':
```

**Comment by lhanneus** [comment]
@mlo-poc did you made a fork ? I have the same issue that you describe.

**Comment by mlo-poc** [comment]
> @mlo-poc did you made a fork ? I have the same issue that you describe.

what I posted here is the full diff.
I diden't fork it, just made my changes and posted the diff.

It's a "works for me" sort of fix.

**Comment by galou** [comment]
Solved in 9395749. Support for Python2 dropped. Users wanting Python2 will have to stick to 389ceac.
